x/custom/model_test.go

```diff
@@ -0,0 +1,180 @@
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/weavetest"
+
+	"github.com/iov-one/tutorial/morm"
+)
+
+func TestValidateCustomStateIndexed(t *testing.T) {
+	now := weave.AsUnixTime(time.Now())
+
+	cases := map[string]struct {
+		model   morm.Model
+		wantErr *errors.Error
+	}{
+		"success, with id": {
+			model: &CustomStateIndexed{
+				Metadata:       &weave.Metadata{Schema: 1},
+				ID:             weavetest.SequenceID(1),
+				CustomString:   "cstm_string",
+				CustomByte:     []byte{0, 1},
+				InnerStateEnum: InnerStateEnum_CaseOne,
+				DeletedAt:      now,
+			},
+			wantErr: nil,
+		},
+		"success, no id": {
+			model: &CustomStateIndexed{
+				Metadata:       &weave.Metadata{Schema: 1},
+				CustomString:   "cstm_string",
+				CustomByte:     []byte{0, 1},
+				InnerStateEnum: InnerStateEnum_CaseOne,
+				DeletedAt:      now,
+			},
+			wantErr: nil,
+		},
+		"failure, missing metadata": {
+			model: &CustomStateIndexed{
+				ID:             weavetest.SequenceID(1),
+				CustomString:   "cstm_string",
+				CustomByte:     []byte{0, 1},
+				InnerStateEnum: InnerStateEnum_CaseOne,
+				DeletedAt:      now,
+			},
+			wantErr: errors.ErrMetadata,
+		},
+		"failure, missing custom string": {
+			model: &CustomStateIndexed{
+				Metadata:       &weave.Metadata{Schema: 1},
+				ID:             weavetest.SequenceID(1),
+				CustomByte:     []byte{0, 1},
+				InnerStateEnum: InnerStateEnum_CaseOne,
+				DeletedAt:      now,
+			},
+			wantErr: errors.ErrEmpty,
+		},
+		"failure, custom string does not begin with 'cstm'": {
+			model: &CustomStateIndexed{
+				Metadata:       &weave.Metadata{Schema: 1},
+				ID:             weavetest.SequenceID(1),
+				CustomString:   "string",
+				CustomByte:     []byte{0, 1},
+				InnerStateEnum: InnerStateEnum_CaseOne,
+				DeletedAt:      now,
+			},
+			wantErr: errors.ErrInput,
+		},
+		"failure, missing inner state enum": {
+			model: &CustomStateIndexed{
+				Metadata:     &weave.Metadata{Schema: 1},
+				ID:           weavetest.SequenceID(1),
+				CustomString: "cstm_string",
+				CustomByte:   []byte{0, 1},
+				DeletedAt:    now,
+			},
+			wantErr: errors.ErrState,
+		},
+		"failure, missing custom byte": {
+			model: &CustomStateIndexed{
+				Metadata:       &weave.Metadata{Schema: 1},
+				ID:             weavetest.SequenceID(1),
+				CustomString:   "cstm_string",
+				InnerStateEnum: InnerStateEnum_CaseOne,
+				DeletedAt:      now,
+			},
+			wantErr: errors.ErrEmpty,
+		},
+		"failure, missing deleted at": {
+			model: &CustomStateIndexed{
+				Metadata:       &weave.Metadata{Schema: 1},
+				ID:             weavetest.SequenceID(1),
+				CustomString:   "cstm_string",
+				CustomByte:     []byte{0, 1},
+				InnerStateEnum: InnerStateEnum_CaseOne,
+			},
+			wantErr: errors.ErrEmpty,
+		},
+	}
+
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			if err := tc.model.Validate(); !tc.wantErr.Is(err) {
+				t.Fatalf("unexpected error: %+v", err)
```

Comment:
You can use [`assert.FieldError`](https://godoc.org/github.com/iov-one/weave/weavetest/assert#FieldError) for testing. Field errors are a new addition and that is why we do not use `FieldError` helper all that much just yet.

With `FieldError` you can assert many errors at once. You can ensure at the same time that errors for certain fields are not reported.  An example use:
- https://github.com/iov-one/weave/blob/49d04133ae28e97517637bfe8686aaa78033e563/x/paychan/msg_test.go
- https://github.com/iov-one/weave/blob/59f4cdd8b7ffb7bd432e52099b63c98b3a0e42c6/x/cron/model_test.go#L64